Fixes for pkgsLLVM with llvmPackages_{15,16,git}

pwaller opened the pull request:
## Description of changes

cc maintainers @dtzWill @ericson2314 @lovek323 @primeos @qyliss @raitobezarius @rrbutani @sternenseemann

With the following overlay, I'm able to build and run `pkgsLLVM.nix` (i.e, nontrivial C++ software) with LLVM 16.

As others have observed, the runtimes build machinery changed quite a bit in recent LLVM versions, so it was a little tricky to make this work, and I've had to tweak the build a bit. I suspect my tweaks may be have the potential to break other platforms, so help wanted!

Overlay:

```nix
import /home/pwaller/.local/src/nixpkgs {
  overlays = [(self: super: {
    llvmPackages = self.llvmPackages_16;

    libseccomp = super.libseccomp.overrideAttrs (final: prev: {
      # Some failing tests, not investigated.
      doCheck = false;
    });

    lowdown = super.lowdown.overrideAttrs (final: prev: {
      # Breaks detection of strlcpy (it gets optimized away during conf test).
      hardeningDisable = [ "fortify" ];
    });

    boost = super.boost.overrideAttrs (final: prev: {
      # Required, otherwise clang can't find the linker.
      # Because boost sets --target=x86_64-pc-linux, and the linker is named
      # x86_64-unknown-linux-gnu-lld.
      NIX_CFLAGS_LINK = prev.NIX_CFLAGS_LINK + " --target=${self.stdenv.hostPlatform.config}";
    });

    db48 = super.db48.overrideAttrs (final: prev: {
      configureFlags = prev.configureFlags
        ++ self.lib.optional (super.db48.stdenv.hostPlatform.useLLVM or false)
          # See https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213
          "CFLAGS=-Wno-error=implicit-function-declaration";
    });

    shadow = super.shadow.overrideAttrs (final: prev: {
      configureFlags = prev.configureFlags
        ++ self.lib.optional (super.shadow.stdenv.hostPlatform.useLLVM or false)
          # See https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213
          "CFLAGS=-Wno-error=implicit-function-declaration";
    });

    nix = super.nix.overrideAttrs (final: prev: {
      # Some tests failing (possibly because they want libgcc_s for pthread cancel?).
      postPatch = (prev.postPatch or "") + ''
        #  rm -r $sourceRoot/tests/{build-remote-content-addressed-fixed.sh,build-remote-input-addressed.sh}
        sed -i '/build-remote-input-addressed.sh/d' tests/local.mk
        sed -i '/build-remote-content-addressed-fixed.sh/d' tests/local.mk
        sed -i '/build-remote-content-addressed-floating.sh/d' tests/local.mk
      '';
    });
  })];
}
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

pwaller committed:
llvmPackages_15, llvmPackages_16: Make the pkgsLLVM.stdenv work

What changed:

* Fixed crtbeginS.o and crtendS.o missing
  (they may or may not be called crt{begin_end},{,_shared}.

* Fixed implicit function declaration causing build errors for various
  builds by supplying -Wno-implicit-function-declaration.

* Fixed __cxxabi_config.h missing, by adding -I${cxxabi}/include/c++/v1
  in the wrapper.

* Fixed libcxx failing to build due to missing libunwind symbols by
  including libunwind as a buildInput, and setting
  -DLIBCXX_ADDITIONAL_LIBRARIES=unwind for stdenv.hostPlatform.useLLVM == true.

* libcxxabi wants to find libunwind at libunwind_shared.so, so symlink
  it there in libunwind.

* llvmPackages_16.libcxxabi: Pass -nostdlib via CMAKE_*_LINKER_FLAGS

  Without this flag, the link of libcxxabi.so tries to pull in libgcc and
  friends, from the clang compiler driver.

* Drop unneeded musl hack patch from libcxx.

* Pass -Wno-error=implicit-function-declaration only to compiler-rt

  See LLVM forum discussion:

  https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213

  In summary, LLVM 16 made implicit function declaration an error. This
  happens a lot in configure scripts which can break things.

* llvmPackages_16: !isDarwin: Supply -DLIBCXX_ABI_USE_LLVM_UNWINDER=On

  Otherwise it fails with various undefined references to _Unwind_*
  functions: (full list: _Unwind_DeleteException _Unwind_GetIP
  _Unwind_GetLanguageSpecificData _Unwind_GetRegionStart
  _Unwind_RaiseException _Unwind_Resume _Unwind_SetGR _Unwind_SetIP).

* 16.libcxxabi: Only pass -nostdlib for useLLVM and Darwin builds

What was tested:

* x86_64-linux, aarch64-linux, the stdenv builds.
  * Additionally I was able to get nix to build, with an overlay to fix
    a couple of minor issues in downstream packages (overlay supplied in
    PR #246577.

* aarch64-darwin fails spuriously in a single LLVM test
  strip-preserve-atime.test checking atime timestamps.

* The same for pkgsLLVM with llvmPackages = llvmPackages_15.

Signed-off-by: Peter Waller <p@pwaller.net>

pwaller commented on pkgs/development/compilers/llvm/16/libcxxabi/default.nix:
```diff
@@ -53,7 +53,7 @@ stdenv.mkDerivation rec {
 
   nativeBuildInputs = [ cmake ninja python3 ];
-  buildInputs = lib.optional (!stdenv.isDarwin && !stdenv.hostPlatform.isWasm) libunwind;
+  propagatedBuildInputs = lib.optional (!stdenv.isDarwin && !stdenv.hostPlatform.isWasm) libunwind;
```
Moved to propagatedBuildInputs, otherwise I get link errors for unwind symbols downstream (e.g. building libcxx).

Artturin replied:
`unwind` is not mentioned in the header file, so it does not appear to be necessary to propagate it. Other builds should explicitly depend on it.

Artturin replied:
Perhaps some changes are needed elsewhere if the unwind is perhaps not bundled any more somewhere?

Artturin replied:
Maybe worth looking in to 
https://github.com/NixOS/nixpkgs/blob/82b98eb23fbdc46226bd3cc0b9632a31361d2090/pkgs/development/compilers/llvm/16/default.nix#L157-L161
https://github.com/NixOS/nixpkgs/blob/82b98eb23fbdc46226bd3cc0b9632a31361d2090/pkgs/development/compilers/llvm/16/default.nix#L242-L263

Im not familiar with the LLVM ecosystem

pwaller replied:
Thanks for the review. I think you're right; I think that this is handled for most things in clangUseLLVM, and where it's missing is in a piece that feeds into the libcxx build, possibly `clangNoLibcxx`? I intend to look into this soon.

pwaller replied:
Removed it and included it directly in libcxx.

reckenrode commented:
I cherry-picked this PR and #241692 onto staging to build and test the Darwin stdenv update.

reckenrode commented on pkgs/development/compilers/llvm/16/default.nix:
```diff
@@ -253,7 +253,18 @@ in let
       nixSupport.cc-cflags =
         [ "-rtlib=compiler-rt"
           "-Wno-unused-command-line-argument"
+          # The compiler got stricter about this, so a number of downstream
+          # things seem to fail with it switched on.
+          "-Wno-implicit-function-declaration"
```
This seems questionable to me. Those things should be fixed. Quite a few have been done already as part of #234710. I still have a few more PRs left to open, but everything in [this gist][1] should build with clang 16. Once #241692 is merged into staging, I expect there will be another round of fixes.

[1]: https://gist.github.com/reckenrode/bf3161814800572e8d38fee52ff3ae40

reckenrode commented on pkgs/development/compilers/llvm/16/default.nix:
```diff
@@ -297,7 +308,12 @@ in let
-      nixSupport.cc-cflags = [ "-nostartfiles" ];
+      nixSupport.cc-cflags = [
+        "-nostartfiles"
+        # The compiler got stricter about this, so a number of downstream
+        # things seem to fail with it switched on.
+        "-Wno-implicit-function-declaration"
```
Same as above. Are there any packages on Linux where you’re encountering problems that haven’t fixed yet on master and staging?

reckenrode commented:
> I've also got an aarch64-darwin build going but it's going to take a long time and the first time around failed in an LLVM test called `strip-preserve-atime` comparing timestamps; in the LLVM tests. It's the only failing LLVM test, I don't believe that's related. So I'm trying another build with `doCheckByDefault = false` to save some time.

That’s interesting. I’ve only encountered a few transient failures on x86_64-darwin while working on the stdenv update. Unfortunately, I can’t remember off hand what they were. In general, the tests should run successfully.

reckenrode commented on pkgs/development/compilers/llvm/16/libcxxabi/default.nix:
```diff
@@ -67,6 +67,7 @@ stdenv.mkDerivation rec {
     "-DCMAKE_CXX_COMPILER_WORKS=ON"
   ] ++ lib.optionals (stdenv.hostPlatform.useLLVM or false) [
     "-DLLVM_ENABLE_LIBCXX=ON"
+  ] ++ lib.optionals (!stdenv.isDarwin) [
```
Darwin isn’t `useLLVM`, so this can continue to be used unconditionally.

pwaller replied:
I removed this per your suggestion. But doing so breaks `pkgs.libcxxabi` (per failing checks marked on the review) Did we both misread this?

Misread:

"The addition is not necessary becuase stdenv.hostPlatform.useLLVM is mutually exclusive with stdenv.isDarwin"

Actual effect / intent:

"All non-darwin platforms gain `-DLIBCXXABI_USE_LLVM_UNWINDER=ON`".

I've reinstated this in cdc598cc33acfcd8ec2de3d35307fe845c62bccc.

(Otherwise they fail with various undefined references to `_Unwind_*` functions: (full list: `_Unwind_DeleteException _Unwind_GetIP _Unwind_GetLanguageSpecificData _Unwind_GetRegionStart _Unwind_RaiseException _Unwind_Resume _Unwind_SetGR _Unwind_SetIP`)

pwaller replied:
I don't think I'm at the bottom of this. I'm seeing something *really* weird. nix-diff shows the only difference between these two drvs is:

```diff

- /nix/store/ihnyk8aph48r3n25sp62iwh3bd91qpgr-libcxxabi-16.0.6.drv:{out}
+ /nix/store/hpl0mrz00kq3pixj0f1n7n41a82j0fh9-libcxxabi-16.0.6.drv:{out}
• The input derivation named `libunwind-16.0.6` differs
  - /nix/store/6w3yyn5fpbnz30v522a2c28ab3dk1yil-libunwind-16.0.6.drv:{dev}
  + /nix/store/6n4mvbg01qwmjn4nlnhqaihg2qlil91m-libunwind-16.0.6.drv:{dev}
  • The environments do not match:
      + postInstall=# libcxxabi wants to link to libunwind_shared.so (?).
ln -s $out/lib/libunwind.so $out/lib/libunwind_shared.so

• Skipping environment comparison

```


But then if I try to build the one with the extra postInstall, I get link errors related to unwind symbols. This makes no sense: these errors are happening before postInstall, and the drv are otherwise identical.

pwaller replied:
Oh, hmm, it seems nix-diff doesn't show environment comparisons by default. I've added `-DCMAKE_EXE_LINKER_FLAGS=-nostdlib -DCMAKE_SHARED_LINKER_FLAGS=-nostdlib`; previously compiler-rt cmake was supplying `-lgcc_s` in order to acquire the unwind symbols, but that's not happening if `-nostdlib` is passed, so I need to change the condition to `((stdenv.hostPlatform.useLLVM or false) || stdenv.isDarwin)`, or leave the darwin bootstrap alone and put the condition in with the other `(stdenv.hostPlatform.useLLVM or false)` parts.

pwaller replied:
d72b64dedd6a056261f0dfa64c4ecd7b96cab85a fixes this, with my first suggestion. I've left the `-nostdlib` in the definition of libcxxabi since it feels more like a libcxxabi issue which could conceivably be fixed some day.

reckenrode commented on pkgs/development/compilers/llvm/16/libcxx/default.nix:
```diff
@@ -62,7 +60,10 @@ stdenv.mkDerivation rec {
-  buildInputs = lib.optionals (!headersOnly) [ cxxabi ];
+  buildInputs = (
+    lib.optional (!headersOnly) cxxabi ++
+    lib.optional ((stdenv.hostPlatform.useLLVM or false) && !stdenv.isDarwin) libunwind
```
Darwin isn’t `useLLVM`, so it doesn’t need to be included in the conditional.

```suggestion
    lib.optional (stdenv.hostPlatform.useLLVM or false) libunwind
```

reckenrode commented on pkgs/development/compilers/llvm/16/libcxx/default.nix:
```diff
@@ -62,7 +60,10 @@ stdenv.mkDerivation rec {
     ++ lib.optional stdenv.isDarwin fixDarwinDylibNames;
 
-  buildInputs = lib.optionals (!headersOnly) [ cxxabi ];
+  buildInputs = (
+    lib.optional (!headersOnly) cxxabi ++
```
Unless the optional is only ever going to be one thing, it seems better to leave it as a list. This also matches the the style used elsewhere in the LLVM derivations.

```suggestion
  buildInputs = lib.optionals (!headersOnly) [ cxxabi ]
```

reckenrode commented on pkgs/development/compilers/llvm/16/default.nix:
```diff
@@ -306,6 +322,13 @@ in let
       bintools = bintools';
       extraPackages = [ ];
       extraBuildCommands = mkExtraBuildCommands0 cc;
+      nixSupport.cc-cflags = [
```
This breaks the Darwin stdenv bootstrap, which needs to use clangNoCompilerRtWithLibc to build Bash, Libsystem, CF, and their dependencies.

```suggestion
      nixSupport.cc-cflags = lib.optionals (!stdenv.hostPlatform.isDarwin) [
```

pwaller replied:
I found a possibly better solution, which is to pass `-nostdlib` only to libcxxabi (which it turns out the darwin bootstrap also used early on, so I removed it from there). This prevents clang from trying to link the startfiles and libgcc.

reckenrode commented:
The aarch64-darwin stdenv built after I made the cc-flags change conditional. I’m building x86_64-darwin now, which I’ll check in the morning.

I deleted my comments regarding `-Wno-implicit-function-declaration` because I missed that it was for a specific clang. I do wonder whether it’s really necessary considering #234710 will require fixing those issues anyway, and many have been fixed already, but since it doesn’t really affect Darwin, I’m less concerned about it.

pwaller reviewed (commented):
Thank for the speedy reviews.

pwaller commented:
> I deleted my comments regarding `-Wno-implicit-function-declaration` because I missed that it was for a specific clang. I do wonder whether it’s really necessary considering #234710 will require fixing those issues anyway, and many have been fixed already, but since it doesn’t really affect Darwin, I’m less concerned about it

I dug into this a little more. Note that there were two places I added it, one affects the build of compiler-rt, and was `clangUseLLVM`. The latter I had to do to get `pkgsLLVM.db48` building, this failed in the configure step with:

```
       > checking for mutexes... UNIX/fcntl
       > configure: WARNING: NO SHARED LATCH IMPLEMENTATION FOUND FOR THIS PLATFORM.
       > configure: error: Unable to find a mutex implementation
```

Where what is going on internally is that the compiler is failing because of implicit function declaration. I suspect there will be a lot of software out there which now fails (in non-obvious ways like the above) with the implicit function declaration issue, and I didn't want to keep bumping into it so I decided to put it in clangUseLLVM for now. Which I admit is not ideal, but finding and fixing all of the 'broken in nonobvious ways' configure scripts which could be affected by this doesn't seem straightforward either.

The compiler-rt build is broken with undeclared functions because nixpkgs patches out the `#include <assert.h>` so we can do a no-libc build. Previously this just gave a warning of implicit function declaration, but in more recent versions of clang this is an error.

What I don't understand is why compiler-rt is including this from libc anyway, since I would expect it to build without libc. So it seems possible this is a compiler-rt bug? I don't see how we can provide an assert.h / assert definition without libc, and I expect that the compiler-rt builtins should work without it in principle, no?

pwaller commented:
The exact error I'm getting from clang by the way is:
```
error: call to undeclared function '<function name here>'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
```

I found discussion on the LLVM forum from Sept/Oct 2022 started by mgorny how it broke Gentoo here:
https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213

The change originally landed in 15.0.0 but was reverted in 15.0.1 and relanded in 16.

mgorny posts about package breakage frequency:

https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213/69

(~4% of packages broken on Gentoo)

It's a bit of a grim situation. Ideally nixpkgs wouldn't put a hack in the compiler, but on the other hand, it's not great to have to hack a large number of packages on a temporary-ish basis either. I don't know if there is precedent about that here.

pwaller commented:
OK: For a moment there I included a patch which touched db and shadow, then I realised that was not a great idea. I've edited the body of the PR and put that in my 'overlay of hacks needed to build nix' instead, motivated by the need to avoid a big rebuild (note this PR is currently targeting master).

I've settled on not making the compiler do `-Wno-error=implicit-function-declaration` by default. Clearly this will need lots of fixes all over nixpkgs over time. grepping around I can see people are already starting to insert it and similar flags in various places.

reckenrode commented:
> > I deleted my comments regarding `-Wno-implicit-function-declaration` because I missed that it was for a specific clang. I do wonder whether it’s really necessary considering #234710 will require fixing those issues anyway, and many have been fixed already, but since it doesn’t really affect Darwin, I’m less concerned about it
> 
> I dug into this a little more. Note that there were two places I added it, one affects the build of compiler-rt, and was `clangUseLLVM`. The latter I had to do to get `pkgsLLVM.db48` building, this failed in the configure step with:
> 
> ```
>        > checking for mutexes... UNIX/fcntl
>        > configure: WARNING: NO SHARED LATCH IMPLEMENTATION FOUND FOR THIS PLATFORM.
>        > configure: error: Unable to find a mutex implementation
> ```

I’d have expected that to have been fixed by #241178 and #242060. I do sometimes test Linux, but I don’t check pkgsLLVM. Fixing things on Darwin is already a fair bit of work. 😅

> Where what is going on internally is that the compiler is failing because of implicit function declaration. I suspect there will be a lot of software out there which now fails (in non-obvious ways like the above) with the implicit function declaration issue, and I didn't want to keep bumping into it so I decided to put it in clangUseLLVM for now. Which I admit is not ideal, but finding and fixing all of the 'broken in nonobvious ways' configure scripts which could be affected by this doesn't seem straightforward either.

I’ve spent the last few months trying to proactively address these before #241692 is merged. Many of these issues are fixed upstream, so I just need to pull the appropriate patches or bump the version in nixpkgs. Where it gets painful is with old stuff that’s still packaged and provided (like Berkley DB, which has the added annoyance you can’t use patches from newer versions due to the licensing changes).

Anyway, as long as this error remains enabled on Darwin by default in the stdenv, it can be suppressed as-needed for pkgsLLVM on Linux (unless Linux maintainers indicate otherwise of course). What I don’t want is for Darwin to deviate by default from upstream when they’ve chosen to promote a particular warning has been promoted to an error.

> The compiler-rt build is broken with undeclared functions because nixpkgs patches out the `#include <assert.h>` so we can do a no-libc build. Previously this just gave a warning of implicit function declaration, but in more recent versions of clang this is an error.
> 
> What I don't understand is why compiler-rt is including this from libc anyway, since I would expect it to build without libc. So it seems possible this is a compiler-rt bug? I don't see how we can provide an assert.h / assert definition without libc, and I expect that the compiler-rt builtins should work without it in principle, no?

That does sound like a bug. I guess this has “worked” in the past because no one has actually tried linking against whichever compiler-rt runtime is using assert on a no-libc build.

[gist]: https://gist.github.com/reckenrode/bf3161814800572e8d38fee52ff3ae40

pwaller commented:
> I’d have expected that to have been fixed by #241178 and #242060. I do sometimes test Linux, but I don’t check pkgsLLVM. Fixing things on Darwin is already a fair bit of work. sweat_smile

Ah cool. I may take a moment to see if I can understand why those fixes aren't working in my case.

> I’ve spent the last few months trying to proactively address these before #241692 is merged. 

Phew, sounds like a lot of work-- thanks for your contributions there!

> That does sound like a bug. I guess this has “worked” in the past because no one has actually tried linking against whichever compiler-rt runtime is using assert on a no-libc build.

You mean a compiler-rt bug?

pwaller commented:
```
I guess this has “worked” in the past because no one has actually tried linking against whichever compiler-rt runtime is using assert on a no-libc build.
```
Well, it would also require compiler-rt to be built with asserts on: it gets defined-away unless you have NDEBUG set. So there is no link-time effect of using assert.

reckenrode commented:
> The exact error I'm getting from clang by the way is:
> 
> ```
> error: call to undeclared function '<function name here>'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
> ```

Invalid function pointer conversions and strict prototypes are the other two big ones I’ve been hitting. There have also been a handful of C++ packages needing fixes because clang 16 defaults to C++17 (which removes the `register` storage class specifier and sometimes wants to use aligned allocations, which is kind of funky on x86_64-darwin due to the SDK).

> I found discussion on the LLVM forum from Sept/Oct 2022 started by mgorny how it broke Gentoo here: https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213
> 
> The change originally landed in 15.0.0 but was reverted in 15.0.1 and relanded in 16.
> 
> mgorny posts about package breakage frequency:
> 
> https://discourse.llvm.org/t/configure-script-breakage-with-the-new-werror-implicit-function-declaration/65213/69
> 
> (~4% of packages broken on Gentoo)
> 
> It's a bit of a grim situation. Ideally nixpkgs wouldn't put a hack in the compiler, but on the other hand, it's not great to have to hack a large number of packages on a temporary-ish basis either. I don't know if there is precedent about that here.

Thanks for the links. When I began working on the clang bump for Darwin, I just started fixing things as they broke. I figured other people would be hitting this problem too (especially now that it looks like Apple will be shipping clang 16 with Xcode 15).  I’m tracking that work in #234710. Obviously one person can’t catch everything, but I hope staging-next is not too much of a mess when the bump is merged.

reckenrode commented:
> > I’d have expected that to have been fixed by #241178 and #242060. I do sometimes test Linux, but I don’t check pkgsLLVM. Fixing things on Darwin is already a fair bit of work. sweat_smile
> 
> Ah cool. I may take a moment to see if I can understand why those fixes aren't working in my case.

It’s possible I didn’t fix the pthread checks completely since Darwin uses its own mutex implementation (which I ended up replacing with `os_unfair_lock_t` to avoid using deprecated/private APIs). db48 was using pthreads on Darwin before the patch, but I fixed it to use the Darwin locking implementation (since that seemed to be upstream’s intention).

> > That does sound like a bug. I guess this has “worked” in the past because no one has actually tried linking against whichever compiler-rt runtime is using assert on a no-libc build.
> 
> You mean a compiler-rt bug?

Yeah, though you make a good point about its being optimized away on release builds. Still, upstream’s build system should do the right thing by default in the no-libc case (assuming it’s an intended/supported build scenario).

pwaller commented:
:+1:, at this point I consider this ready to go (pending ofbor checks, and any further tests you want to do) unless someone can spot any further issues with it. I'm not currently trying to do any darwin tests.

reckenrode commented on pkgs/development/compilers/llvm/16/default.nix:
```diff
@@ -323,6 +315,7 @@ in let
       extraPackages = [ ];
       extraBuildCommands = mkExtraBuildCommands0 cc;
     };
+
```
```suggestion
```

reckenrode commented on pkgs/development/compilers/llvm/16/libcxxabi/default.nix:
```diff
@@ -65,9 +65,16 @@ stdenv.mkDerivation rec {
   ] ++ lib.optionals (stdenv.hostPlatform.useLLVM or false) [
     "-DLLVM_ENABLE_LIBCXX=ON"
     "-DLIBCXXABI_USE_LLVM_UNWINDER=ON"
+  ] ++ lib.optionals ((stdenv.hostPlatform.useLLVM or false) || stdenv.isDarwin) [
```
```suggestion
  ] ++ lib.optionals ((stdenv.hostPlatform.useLLVM or false) || stdenv.hostPlatform.isDarwin) [
```

reckenrode commented on pkgs/development/compilers/llvm/16/libcxxabi/default.nix:
```diff
@@ -65,9 +65,16 @@ stdenv.mkDerivation rec {
     # CMake however checks for this anyways; this flag tells it not to. See:
     # https://github.com/llvm/llvm-project/blob/4bd3f3759259548e159aeba5c76efb9a0864e6fa/llvm/runtimes/CMakeLists.txt#L243
     "-DCMAKE_CXX_COMPILER_WORKS=ON"
+
```
```suggestion
```

pwaller commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -697,7 +697,6 @@ in
           };
 
           libcxxBootstrapStdenv = self.overrideCC stdenvNoCF (self.llvmPackages.clangNoCompilerRtWithLibc.override {
-            nixSupport.cc-cflags = [ "-nostdlib" ];
```
Good point.

reckenrode commented:
> 👍, at this point I consider this ready to go (pending ofbor checks, and any further tests you want to do) unless someone can spot any further issues with it. I'm not currently trying to do any darwin tests.

I left a few Darwin-related comments. The stdenv changes should be dropped, but it otherwise looks good. I’ll defer to the LLVM maintainers, but I expect the working commits will need squashed.

pwaller commented:
Thanks for the quick reviews. Addressd and squashed.

reckenrode reviewed (approved):
Thanks! LGTM, but it’d definitely be good to get a review from one of the LLVM maintainers.

reckenrode commented on pkgs/development/compilers/llvm/16/libcxxabi/default.nix:
```diff
@@ -68,6 +68,12 @@ stdenv.mkDerivation rec {
   ] ++ lib.optionals (stdenv.hostPlatform.useLLVM or false) [
     "-DLLVM_ENABLE_LIBCXX=ON"
     "-DLIBCXXABI_USE_LLVM_UNWINDER=ON"
+  ] ++ lib.optionals ((stdenv.hostPlatform.useLLVM or false) || stdenv.hostPlatform.isDarwin) [
```
I opened #247066 to fix pkgsStatic on Darwin with LLVM 16. `enableShared` needs to be added to the condition, or the static libc++abi will fail to build.

```suggestion
  ] ++ lib.optionals ((stdenv.hostPlatform.useLLVM or false) || (stdenv.hostPlatform.isDarwin && enableShared) [
```

pwaller replied:
Done

reckenrode commented on pkgs/development/compilers/llvm/16/libunwind/default.nix:
```diff
@@ -40,6 +40,11 @@ stdenv.mkDerivation rec {
     cd ../runtimes
   '';
 
+  postInstall = lib.optionalString (enableShared) ''
+    # libcxxabi wants to link to libunwind_shared.so (?).
+    ln -s $out/lib/libunwind.so $out/lib/libunwind_shared.so
+  '';
+
```
I didn’t think this would really affect Darwin since it uses Apple’s implementation of libunwind, but then I noticed recently that Apple dropped their implementation in favor of shipping the one from LLVM (since macOS 11). Making this conditional would allow nixpkgs to also use LLVM’s libunwind on Darwin in the future.

pwaller replied:
Condition now reads:
```nix
postInstall = lib.optionalString (enableShared && !stdenv.hostPlatform.isDarwin) ''
```

tobim reviewed (approved):
This PR allows me to build a statically linked `hello` with clang16 with the following extra modifications:

```diff
diff --git a/pkgs/stdenv/adapters.nix b/pkgs/stdenv/adapters.nix
index f29bdf671c8c..9c8ff82e1ec4 100644
--- a/pkgs/stdenv/adapters.nix
+++ b/pkgs/stdenv/adapters.nix
@@ -61,7 +61,8 @@ rec {
       if stdenv.hostPlatform.isDarwin
       then throw "Cannot build fully static binaries on Darwin/macOS"
       else (mkDerivationSuper args).overrideAttrs(finalAttrs: {
-        NIX_CFLAGS_LINK = toString (finalAttrs.NIX_CFLAGS_LINK or "") + " -static";
+        NIX_CFLAGS_LINK = toString (finalAttrs.NIX_CFLAGS_LINK or "") + " -static -Wl,--no-dynamic-linker";
       } // lib.optionalAttrs (!(finalAttrs.dontAddStaticConfigureFlags or false)) {
         configureFlags = (finalAttrs.configureFlags or []) ++ [
             "--disable-shared" # brrr...
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 3b27655c631b..4623e93080df 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -16306,7 +16306,7 @@ with pkgs;
       /**/ if platform.isDarwin then 11
       else if platform.isFreeBSD then 12
       else if platform.isAndroid then 12
-      else if platform.isLinux then 11
+      else if platform.isLinux then 16
       else if platform.isWasm then 12
       else 14;
     # We take the "max of the mins". Why? Since those are lower bounds of the
diff --git a/pkgs/top-level/stage.nix b/pkgs/top-level/stage.nix
index 3886ae04e492..65dda7012363 100644
--- a/pkgs/top-level/stage.nix
+++ b/pkgs/top-level/stage.nix
@@ -269,6 +269,24 @@ let
         gcc.abi = "elfv2";
       };
     });
+
+    # Fully static packages built with clang.
+    # Currently uses Musl on Linux (couldn’t get static glibc to work).
+    pkgsStaticLLVM = nixpkgsFun ({
+      overlays = [ (self': super': {
+        pkgsStaticLLVM = super';
+      })] ++ overlays;
+      crossSystem = {
+        useLLVM = true;
+        isStatic = true;
+        parsed =
+          if stdenv.isLinux
+          then makeMuslParsedPlatform stdenv.hostPlatform.parsed
+          else stdenv.hostPlatform.parsed;
+      } // lib.optionalAttrs (stdenv.hostPlatform.system == "powerpc64-linux") {
+        gcc.abi = "elfv2";
+      };
+    });
   };
 
   # The complete chain of package set builders, applied from top to bottom.
```

That is pretty great!

RaitoBezarius commented:
@pwaller Are those changes unique to LLVM16 or do they make sense for:

(a) LLVM git
(b) LLVM 15?

pwaller commented:
They may. I can investigate. Do you have an opinion on how it should affect this pull request? Should this one handle all of them at once?

pwaller commented:
`pkgsLLVM(15).llvmPackages.compiler-rt-no-libc`'s stdenv doesn't have an $AR (and other variables) set in it. I can't immediately see why. Diffing various derivations doesn't show any significant differences which might explain this (stdenv, stdenv.cc, .cc.bintools, for example). This is the first package in the chain and it doesn't compile because CMAKE_AR is set to an empty string, so the build step to build .a files results in ninja trying to exec an empty string.

pwaller commented:
@RaitoBezarius I wasn't able to get time to dig into this for now, unfortunately I had store corruption which manifested in an awkward way and wasn't quickly revealed. I'll be back next week. Is it possible to consider accepting the improvements to LLVM 16 and then addressing LLVM 15 and git separately? I'm interested in seeing those work well as well, it's just a matter of getting time for it. My hope is that if key ideas are present on master for LLVM 16 at least, then those fixes can make their way to the other versions in time.

pwaller commented:
Rebased against current master and applied relevant fixes for llvmPackages_15, which allows them to build. `pkgsLLVM.hello` also builds. However I'm not able to build `pkgsLLVM.nix` (with the overlay) because I get:
```
       > x86_64-unknown-linux-gnu-ld: error: unable to find library -l
```
I haven't yet dug into this further, but I would prefer not to hold up landing this; those other issues can be found and fixed by others wanting to experiment with this toolchain.

I haven't yet looked at llvmPackages_git.

pwaller commented:
I just tried llvmPackages_git as well but that seems broken in new ways I'm not willing to dig into in the scope of this PR. Maybe later?

<details>
<summary>llvmPackages_git issues</summary>
```
error:
       … while evaluating the attribute 'stdenv' of the derivation 'hello-x86_64-unknown-linux-gnu-2.12.1'

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:300:7:

          299|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          300|       name =
             |       ^
          301|         let

       … while evaluating the attribute 'defaultNativeBuildInputs' of the derivation 'stdenv-linux'

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/default.nix:96:14:

           95|     // {
           96|       inherit name;
             |              ^
           97|       inherit disallowedRequisites;

       … while evaluating the attribute 'depsTargetTargetPropagated' of the derivation 'x86_64-unknown-linux-gnu-clang-wrapper-15.0.7'

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:300:7:

          299|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          300|       name =
             |       ^
          301|         let

       … while evaluating the attribute 'buildInputs' of the derivation 'libcxx-x86_64-unknown-linux-gnu-15.0.7'

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:300:7:

          299|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          300|       name =
             |       ^
          301|         let

       … while evaluating the attribute 'cmakeFlags' of the derivation 'libcxxabi-x86_64-unknown-linux-gnu-15.0.7'

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:300:7:

          299|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          300|       name =
             |       ^
          301|         let

       … while evaluating the attribute 'stdenv' of the derivation 'libcxx-headers-x86_64-unknown-linux-gnu-15.0.7'

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:300:7:

          299|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          300|       name =
             |       ^
          301|         let

       error: infinite recursion encountered

       at /home/pwaller/.local/src/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:329:14:

          328|       args = attrs.args or ["-e" (attrs.builder or ./default-builder.sh)];
          329|       inherit stdenv;
             |              ^
```
</details>

pwaller committed:
llvmPackages_git: Port stdenv fix for cxx-headers

Signed-off-by: Peter Waller <p@pwaller.net>

pwaller commented:
@RaitoBezarius please take another look, I've addressed your question.

> @pwaller Are those changes unique to LLVM16 or do they make sense for:
> 
> (a) LLVM git (b) LLVM 15?

I've applied the changes to LLVM git and LLVM. `pkgsLLVM.hello` builds and runs for both of those now. Some downstream C++ stuff is broken which I have not investigated since it works for LLVM 16.

pwaller commented:
badd8c3cb8ecf9a99465aac53b40b353f25a8e4a fixed the infinite recursion issue I hit earlier. The git log suggests that you added that fix to LLVM 16 when it landed.

RaitoBezarius commented on pkgs/development/compilers/llvm/15/libcxx/default.nix:
```diff
@@ -75,8 +77,18 @@ stdenv.mkDerivation rec {
+      # does not appear to)
+      # There's precedent for this in llvm-project/libcxx/cmake/caches.
+      # In a monorepo build you might do the following in the libcxxabi build:
+      #   -DLLVM_ENABLE_PROJECTS=libcxxabi;libunwinder
```
```suggestion
      #   -DLLVM_ENABLE_PROJECTS=libcxxabi;libunwind
```

RaitoBezarius replied:
or is it really unwinder?

pwaller replied:
No, it's not, typo must have slipped in at some point.

RaitoBezarius commented:
OK, I checked a nixpkgs-review and this builds.
Please can you rebase? I will approve and we can move forward.

Ericson2314 commented on pkgs/development/compilers/llvm/15/libcxxabi/default.nix:
```diff
@@ -68,6 +68,13 @@ stdenv.mkDerivation rec {
     "-DLIBCXXABI_USE_LLVM_UNWINDER=ON"
+  ] ++ lib.optionals ((stdenv.hostPlatform.useLLVM or false) ||
+                      (stdenv.hostPlatform.isDarwin && enableShared)) [
+    # libcxxabi's CMake looks as though it treats -nostdlib++ as implying -nostdlib,
+    # but that does not appear to be the case for example when building
+    # pkgsLLVM.libcxxabi (which uses clangNoCompilerRtWithLibc).
+    "-DCMAKE_EXE_LINKER_FLAGS=-nostdlib"
+    "-DCMAKE_SHARED_LINKER_FLAGS=-nostdlib"
```
What is the reason for this? I would think we *do* wanna link libc but *don't* want to link (an already-existing) C++ standard library implementation. `-nostdlib++` without `-nostdlib` therefore seems correct to me.

pwaller replied:
Apologies, this is coming back to it some time after I implemented it, so I have forgotten the detail here and I have limited cycles to spare to dive back in. I suspect it has to do with making cmake configure, or linking succeed. I presume if you take it out you'll find the same failure I experienced? If not, maybe there is something further downstream which is broken.

Ericson2314 replied:
OK thanks @pwaller. So far, it actually seems to be working fine for me.

rrbutani replied:
@pwaller: I think it's (1); see: https://github.com/NixOS/nixpkgs/pull/311836/commits/c84a96bd96ae719a6f9b679e84424c86dbf4b6cf from #311836

pwaller replied:
Thanks for the cross reference. I looked into it a little, I suspect the issue this was trying to fix went away when libcxxabi was merged into libcxx.
